### PR TITLE
Added Redis based stores (security & clients)

### PR DESCRIPTION
**Pull request by jvermillard**
CQ under review for jedis + common-pool2
http://dev.eclipse.org/ipzilla/show_bug.cgi?id=11027
http://dev.eclipse.org/ipzilla/show_bug.cgi?id=11029
so don't merge before we get the go :)

However I would appreciate feedback

### CODE REVIEW

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            byte[] k = (EP_CLIENT + client.getEndpoint()).getBytes();
+            byte[] old = j.getSet(k, data);
+
+            j.expire(k, client.getLifeTimeInSec().intValue());
+
+            // secondary index
+            byte[] idx = (REG_EP + client.getRegistrationId()).getBytes();
+            j.set(idx, client.getEndpoint().getBytes());
+
+            j.expire(idx, client.getLifeTimeInSec().intValue());
```
Comment (msangoi) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java]: No call to ClientRegistryListener.unregistered() when the lifetime is expired.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        }
+    }
+
+    @Override
+    public Client deregisterClient(String registrationId) {
+        try (Jedis j = pool.getResource()) {
+            // first rename for atomicity
+            if (!"OK".equals(j.rename((REG_EP + registrationId).getBytes(),
+                    ("TODELETE#" + REG_EP + registrationId).getBytes()))) {
+                System.err.println("del meh");
```
Comment (msangoi) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java]: Syserr meh

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            byte[] key = j.get(("TODELETE#" + REG_EP + registrationId).getBytes());
+            if (key == null) {
+                return null;
+            }
+            byte[] data = j.get(ByteArrayUtils.concatenate(EP_CLIENT.getBytes(), key));
+
+            Client c = deserialize(data);
+
+            // delete everything
+            j.del(("TODELETE#" + REG_EP + registrationId).getBytes());
```
Comment (msangoi) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java]: `("TODELETE#" + REG_EP + registrationId).getBytes() ` -> This key is built 3 times in this method.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            j.del(ByteArrayUtils.concatenate(EP_CLIENT.getBytes(), key));
```
Comment (msangoi) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java]: This key construction is duplicated too.

**Comment by sophokles73** [comment]
This looks good, @jvermillard :-)
My only concern would be using standard Java serializaiton for marshalling/unmarshalling the objects. The Serializable contract is not easy to implement (and keep for that matter) as e.g. discussed in chapter 11 of Joshua Bloch's *Effective Java*. The arguments Joshua brings forward may not be applicable here (mainly due to the short-lived-ness of the cached objects) but I think it would still be wise to reflect on this a little more and maybe use another (custom) serialization format or use e.g. Jackson to marshall to JSON instead.

**Comment by jvermillard** [comment]
yes java serialization is really for the PoC.
My next move on tha would be to use https://capnproto.org/ 
or to split the objects in redis HSET fields for having the capability to patch them on updates

**Comment by sophokles73** [comment]
The latter sounds reasonable to me. Do you think it is worth the effort of defining a schema and compiling type classes using something like capnproto for the few times we actually serialize, deserialize the objects?

**Comment by jvermillard** [comment]
Now with json serialisation

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java (hunk 1 above)
Comment (jvermillard) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java]: see the TODO is javadoc comment, the plan is to receive expiration from redis

**Comment by jvermillard** [comment]
waiting for CQ then it can be merged (if no other comments)

**Comment by ingoschaal** [comment]
I try to use the new Redis based stores with a redis cloud service for example RedisCloud of pivotal. This service requires authentication.
There are no options, to enter credentials (password). 
Thereby I get this exception redis.clients.jedis.exceptions.JedisDataException: NOAUTH Authentication required.
The Jedis Object provides a mehthode auth(Stirng password) with which it works
Is it provided that it comes with authentication?

**Comment by sophokles73** [comment]
@ingoschaal 

looks like a good opportunity for you to create a PR adding support for authentication ;-) But you should wait until Julien has merged this PR ...

**Comment by jvermillard** [comment]
you can PR on this branch, we are waiting to have observation working for merging, so we wait for https://github.com/eclipse/californium/pull/16 to be done and released

**Comment by ingoschaal** [comment]
Ok, good suggestion!
Another question:
Is it provided that the Redis based implementation of the Clientregistry remains in the demo server or should this be included in the core?
In my opinion, the core is the better place. Maybee in future ?

**Comment by jvermillard** [comment]
IMO the future is to make demo a real deployable server, and keep core as a pure LWM2M protocol library but we can discuss that :) my plan was to open this topic once leshan-demo is working as a cluster and so more than a simple demo server but something nearly deployable out of the box

**Comment by sbernard31** [comment]
We decide to merge #93  and #89 in #113.

**Comment by Fakhruddin-Kararawala** [comment]
@jvermillard 
how can we observe a resource with lt,gt,pmin,pmax parameter as threshold.
what we are trying is that the device can send us notification only when these threshold values exceeds.

**Pull request closed without merging**